packages/hebao_v1/contracts/modules/dapps/CompoundModule.sol

```diff
@@ -133,7 +218,209 @@ contract CompoundModule is BaseSubAccount, SecurityModule
+            uint leftLiquidityToToken
+        )
+    {
+        address cToken = compoundRegistry.getCToken(token);
+        if (cToken == address(0)) {
+            return (0, 0);
+        }
+
+        require(isCollaterlCToken(wallet, token), "NOT_COLLATERAL_CTOKEN");
+
+        address priceOracle = address(Comptroller(comptroller).oracle);
+        uint price = ComptrollerPriceOracle(priceOracle).getUnderlyingPrice(CToken(cToken));
+        require(price != 0, "TOKEN_NO_PRICE");
+
+        (uint err, uint liquidity, ) = Comptroller(comptroller).getAccountLiquidity(wallet);
+        if (err != 0 || liquidity == 0) {
+            return (0, 0);
+        }
+
+        // mapping(address => Comptroller.Market) storage markets = Comptroller(comptroller).markets;
+        (bool isListed, uint collateralFactorMantissa) = ComptrollerV2Storage(comptroller).markets(cToken);
+        if (isListed) {
+            return (0, 0);
+        }
+
+        // calculate current cToken liquidity
+        uint cTokenSupply;
+        uint exchangeRateMantissa;
+        (err, cTokenSupply, , exchangeRateMantissa) = CToken(cToken).getAccountSnapshot(wallet);
+        if (err != 0) {
+            return (0, 0);
+        }
+        uint cTokenToTokenAmount = cTokenSupply.mul(exchangeRateMantissa) / (10 ** 18);
+        uint normalizedTokenLiquidity = cTokenToTokenAmount.mul(price);
+        uint currentHoldLiquidity = normalizedTokenLiquidity.mul(collateralFactorMantissa) / (10 ** 18);
+
+        if (currentHoldLiquidity <= liquidity) {
+            // current holding liquidity <= free liquidity, means we have other enough collaterals.
+            // we can remove this cToken all.
+            safeLiquidityToToken = cTokenToTokenAmount;
+
+            // After removal, we still have free liquidity, so we can borrow more.
+            leftLiquidityToToken = (liquidity - currentHoldLiquidity) / price;
+        } else {
+            // current holding liquidity >= free liquidity, means part of our cToken is already used as collateral.
+            // So we can remove only a part of this cToken.
+            safeLiquidityToToken = liquidity / price;
+            // After removal, we don't have free liquidity, so we can borrow 0.
+            leftLiquidityToToken = 0;
+        }
+    }
+
+    // Compound specific functions
+
+    /// @dev add more collateral to get more withdrawables.
+    function addCollaterals (
+        address          wallet,
+        address[] memory tokens,
+        uint[]    memory amounts
+        )
+        public
+        nonReentrant
+        onlyFromMetaTxOrWalletOwner(wallet)
+        returns (uint)
+    {
+        require(tokens.length > 0, "EMPTY_TOKEN_ARRAY");
+
+        // address[] memory cTokens = new address[](tokens.length);
+        for (uint i = 0; i < tokens.length; ++i) {
+            address cToken = compoundRegistry.getCToken(tokens[i]);
+            if (cToken != address(0) && amounts[i] != 0) {
```

Comment:
I think we should allow `amounts[i] == 0` so that `enterMarketsIfNeeded` is called for the token. Otherwise if the user first used `deposit` to get cTokens there would otherwise be no way to use that existing balance as collateral without redeeming and calling `addCollaterals` again with the redeemed tokens, which is much more inefficient.